libfreerdp/cache/persistent.c

```diff
@@ -0,0 +1,333 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * Persistent Bitmap Cache
+ *
+ * Copyright 2016 Marc-Andre Moreau <marcandre.moreau@gmail.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <winpr/crt.h>
+#include <winpr/stream.h>
+
+#include <freerdp/freerdp.h>
+#include <freerdp/constants.h>
+
+#include <freerdp/log.h>
+#include <freerdp/cache/persistent.h>
+
+#define TAG FREERDP_TAG("cache.persistent")
+
+int persistent_cache_get_version(rdpPersistentCache* persistent)
+{
+	return persistent->version;
+}
+
+int persistent_cache_get_count(rdpPersistentCache* persistent)
+{
+	return persistent->count;
+}
+
+int persistent_cache_read_entry_v2(rdpPersistentCache* persistent, PERSISTENT_CACHE_ENTRY* entry)
+{
+	PERSISTENT_CACHE_ENTRY_V2 entry2;
+
+	if (fread((void*)&entry2, sizeof(PERSISTENT_CACHE_ENTRY_V2), 1, persistent->fp) != 1)
+		return -1;
+
+	entry->key64 = entry2.key64;
+	entry->width = entry2.width;
+	entry->height = entry2.height;
+	entry->size = entry2.width * entry2.height * 4;
+	entry->flags = entry2.flags;
+
+	entry->data = persistent->bmpData;
+
+	if (fread((void*)entry->data, 0x4000, 1, persistent->fp) != 0)
+		return -1;
```

Comment:
Looks like the test of `fread()` result is inverted

Reply:
good catch, I also modified all the fread/fwrite calls in persistent.c to swap the "number" and "count" fields such that the return value is the number of bytes read/written.